## Android vs iOS

In [1]:
#importing required packages and libraries
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

### Android Dataset

In [2]:
#importing andriod dataset
df_1 = pd.read_csv('android.csv')
df_1.head()

App Name  \
0                                            Gakondo   
1                                Ampere Battery Info   
2                                             Vibook   
3  Smart City Trichy Public Service Vehicles 17UC...   
4                                            GROW.me   

                       App Id       Category  Rating  Rating Count Installs  \
0         com.ishakwe.gakondo      Adventure     0.0           0.0      10+   
1  com.webserveis.batteryinfo          Tools     4.4          64.0   5,000+   
2        com.doantiepvien.crm   Productivity     0.0           0.0      50+   
3     cst.stJoseph.ug17ucs548  Communication     5.0           5.0      10+   
4        com.horodyski.grower          Tools     0.0           0.0     100+   

   Minimum Installs  Maximum Installs  Free  Price  ...  \
0              10.0                15  True    0.0  ...   
1            5000.0              7662  True    0.0  ...   
2              50.0                58  True    0.0  ...   
3              10.0                19  True    0.0  ...   
4             100.0               478  True    0.0  ...   

                  Developer Website              Developer Email  \
0       https://beniyizibyose.tk/#/       jean21101999@gmail.com   
1   https://webserveis.netlify.app/         webserveis@gmail.com   
2                               NaN          vnacrewit@gmail.com   
3  http://www.climatesmarttech.com/  climatesmarttech2@gmail.com   
4       http://www.horodyski.com.pl    rmilekhorodyski@gmail.com   

       Released  Last Updated Content Rating  \
0  Feb 26, 2020  Feb 26, 2020       Everyone   
1  May 21, 2020  May 06, 2021       Everyone   
2   Aug 9, 2019  Aug 19, 2019       Everyone   
3  Sep 10, 2018  Oct 13, 2018       Everyone   
4  Feb 21, 2020  Nov 12, 2018       Everyone   

                                      Privacy Policy Ad Supported  \
0                 https://beniyizibyose.tk/projects/        False   
1  https://dev4phones.wordpress.com/licencia-de-uso/         True   
2  https://www.vietnamairlines.com/vn/en/terms-an...        False   
3                                                NaN         True   
4                        http://www.horodyski.com.pl        False   

  In App Purchases Editors Choice         Scraped Time  
0            False          False  2021-06-15 20:19:35  
1            False          False  2021-06-15 20:19:35  
2            False          False  2021-06-15 20:19:35  
3            False          False  2021-06-15 20:19:35  
4            False          False  2021-06-15 20:19:35  

[5 rows x 24 columns]

### Description of Android Dataset:

We are using a csv file "android.csv". Before getting into actual EDA, we will familiarize with the variables :-

* **App Name** - Name of the App
* **App Id** - Id of the APP
* **Category** - Category under which the App falls.
* **Rating count** - Application's rating on playstore
* **Installs** - Number of Installs of the App
* **Minimum Installs** - Maximum Installation Count of the App
* **Maximum Installs** - Minimum Installation Count of the App
* **Free** - Boolean (True/False)
* **Price** - Price of the app (0 if it is Free)
* **Reviews** - Number of reviews of the App.
* **Size** - Size of the App.
* **Minimum Android** - Supported Android Version
* **Developer Id** - ID of the Developer
* **Developer Website** - Developer webite
* **Developer Email** - Developer email is mentioned
* **Released** - App Release Date
* **Last Updated** - Date when the App was last updated
* **Content Rating** - Appropiate Target Audience of the App
* **Privacy Policy** - Link for the Privacy Policy Agreement
* **Ad Supported** - Is App is Ad Supported (Boolean)
* **In App Purchases** - Is App has in app purchases (Boolean)
* **Editors Choice** - Is app showed in Editors Choice (Boolean)
* **Scraped Time** - Data Scraped Time


In [3]:
# Displaying attributes in  android dataset


df_1.columns

Index(['App Name', 'App Id', 'Category', 'Rating', 'Rating Count', 'Installs',
       'Minimum Installs', 'Maximum Installs', 'Free', 'Price', 'Currency',
       'Size', 'Minimum Android', 'Developer Id', 'Developer Website',
       'Developer Email', 'Released', 'Last Updated', 'Content Rating',
       'Privacy Policy', 'Ad Supported', 'In App Purchases', 'Editors Choice',
       'Scraped Time'],
      dtype='object')

In [4]:
# Displaying the  schema summary of a dataframe

df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 799999 entries, 0 to 799998
Data columns (total 24 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   App Name           799998 non-null  object 
 1   App Id             799999 non-null  object 
 2   Category           799999 non-null  object 
 3   Rating             792086 non-null  float64
 4   Rating Count       792086 non-null  float64
 5   Installs           799963 non-null  object 
 6   Minimum Installs   799963 non-null  float64
 7   Maximum Installs   799999 non-null  int64  
 8   Free               799999 non-null  bool   
 9   Price              799999 non-null  float64
 10  Currency           799947 non-null  object 
 11  Size               799932 non-null  object 
 12  Minimum Android    797783 non-null  object 
 13  Developer Id       799991 non-null  object 
 14  Developer Website  536705 non-null  object 
 15  Developer Email    799992 non-null  object 
 16  Re

In [5]:
# Displaying statistics summary of dataset’s distribution with rows and columns, including NaN values.

df_1.describe(include='all')

App Name               App Id   Category         Rating  \
count        799998               799999     799999  792086.000000   
unique       773551               799999         48            NaN   
top     Tic Tac Toe  com.ishakwe.gakondo  Education            NaN   
freq            136                    1      83393            NaN   
mean            NaN                  NaN        NaN       2.203589   
std             NaN                  NaN        NaN       2.106622   
min             NaN                  NaN        NaN       0.000000   
25%             NaN                  NaN        NaN       0.000000   
50%             NaN                  NaN        NaN       2.900000   
75%             NaN                  NaN        NaN       4.300000   
max             NaN                  NaN        NaN       5.000000   

        Rating Count Installs  Minimum Installs  Maximum Installs    Free  \
count   7.920860e+05   799963      7.999630e+05      7.999990e+05  799999   
unique           NaN       21               NaN               NaN       2   
top              NaN     100+               NaN               NaN    True   
freq             NaN   153297               NaN               NaN  784344   
mean    2.910340e+03      NaN      1.644477e+05      2.874675e+05     NaN   
std     2.577076e+05      NaN      9.982313e+06      1.492685e+07     NaN   
min     0.000000e+00      NaN      0.000000e+00      0.000000e+00     NaN   
25%     0.000000e+00      NaN      5.000000e+01      8.400000e+01     NaN   
50%     6.000000e+00      NaN      5.000000e+02      6.940000e+02     NaN   
75%     4.200000e+01      NaN      5.000000e+03      7.345000e+03     NaN   
max     1.385576e+08      NaN      5.000000e+09      6.265638e+09     NaN   

                Price  ...          Developer Website       Developer Email  \
count   799999.000000  ...                     536705                799992   
unique            NaN  ...                     329266                418699   
top               NaN  ...  http://www.subsplash.com/  support@classplus.co   
freq              NaN  ...                       2621                  3564   
mean         0.101888  ...                        NaN                   NaN   
std          2.574308  ...                        NaN                   NaN   
min          0.000000  ...                        NaN                   NaN   
25%          0.000000  ...                        NaN                   NaN   
50%          0.000000  ...                        NaN                   NaN   
75%          0.000000  ...                        NaN                   NaN   
max        400.000000  ...                        NaN                   NaN   

            Released  Last Updated Content Rating  \
count         775559        799999         799999   
unique          4137          3642              6   
top     Feb 19, 2020  Jun 10, 2021       Everyone   
freq             725          3785         699440   
mean             NaN           NaN            NaN   
std              NaN           NaN            NaN   
min              NaN           NaN            NaN   
25%              NaN           NaN            NaN   
50%              NaN           NaN            NaN   
75%              NaN           NaN            NaN   
max              NaN           NaN            NaN   

                                Privacy Policy Ad Supported In App Purchases  \
count                                   654608       799999           799999   
unique                                  387570            2                2   
top     http://www.subsplash.com/legal/privacy        False            False   
freq                                      2594       401284           732265   
mean                                       NaN          NaN              NaN   
std                                        NaN          NaN              NaN   
min                                        NaN          NaN              NaN   
25%                

In [6]:
# Dropping the unwanted columns
df_1.drop("App Id",axis=1, inplace=True)
df_1.drop("Minimum Installs",axis=1, inplace=True)
df_1.drop("Maximum Installs",axis=1, inplace=True)
#df_1.drop("Free",axis=1, inplace=True)
df_1.drop("Currency",axis=1, inplace=True)
df_1.drop("Minimum Android",axis=1, inplace=True)
df_1.drop("Developer Id",axis=1, inplace=True)#
df_1.drop("Developer Website",axis=1, inplace=True)
df_1.drop("Developer Email",axis=1, inplace=True)
df_1.drop("Released",axis=1, inplace=True)
df_1.drop("Last Updated",axis=1, inplace=True)
#df_1.drop("Content Rating",axis=1, inplace=True)
df_1.drop("Privacy Policy",axis=1, inplace=True)
df_1.drop("Ad Supported",axis=1, inplace=True)
df_1.drop("In App Purchases",axis=1, inplace=True)
df_1.drop("Editors Choice",axis=1, inplace=True)
df_1.drop("Scraped Time",axis=1, inplace=True)
df_1.drop("Rating Count",axis=1, inplace=True)

In [7]:
# Displaying to get the first 4 rows

df_1.head()

App Name       Category  Rating  \
0                                            Gakondo      Adventure     0.0   
1                                Ampere Battery Info          Tools     4.4   
2                                             Vibook   Productivity     0.0   
3  Smart City Trichy Public Service Vehicles 17UC...  Communication     5.0   
4                                            GROW.me          Tools     0.0   

  Installs  Free  Price  Size Content Rating  
0      10+  True    0.0   10M       Everyone  
1   5,000+  True    0.0  2.9M       Everyone  
2      50+  True    0.0  3.7M       Everyone  
3      10+  True    0.0  1.8M       Everyone  
4     100+  True    0.0  6.2M       Everyone

In [8]:
#Checking for Null values in the data set
df_1.isnull().sum()

App Name             1
Category             0
Rating            7913
Installs            36
Free                 0
Price                0
Size                67
Content Rating       0
dtype: int64

We can see that there are many NaN (missing) values in our dataset, especially in the Rating and rating Count column.

We can see that, almost 7913 of the rows have null values in place of Ratings. Hence we are taking the mean of the Rating column and filling up the null values.

In case of other variables like, Installs, Size etc, the number of null values are too negligible that one can either drop it or fill it. I am just using a fillna command with forward fill to remove null values in the rest of the columns.

In [9]:
#Rating column
df_1['Rating'].fillna(df_1['Rating'].mean(),inplace=True)

#App Name
df_1['App Name'].fillna(method='ffill',inplace=True)

#Size
df_1['Size'].fillna(method='ffill',inplace=True)

#Installs
df_1['Installs'].fillna(method='ffill',inplace=True)

In [10]:
#Checking for Null values in the data set
df_1.isnull().sum()

App Name          0
Category          0
Rating            0
Installs          0
Free              0
Price             0
Size              0
Content Rating    0
dtype: int64

In [11]:
# Verifying the Maximum count in Rating attribute

df_1['Rating'].max()

5.0

In [12]:
# Removing the special characters from the Install column 
df_1['Install1'] = df_1["Installs"].str.replace('+','')
df_1['Install']=df_1['Install1'].str.replace(',','')
df_1.Size = [x.strip().replace('M', '') for x in df_1.Size]
df_1.Size = [x.strip().replace('k', '') for x in df_1.Size]

/var/folders/q5/92dy6jw11_l3t652x65zqf580000gn/T/ipykernel_61007/1840719356.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_1['Install1'] = df_1["Installs"].str.replace('+','')


In [ ]:
## Changing few categorical to numerical

df_1["Rating"]= df_1.Rating.apply(lambda x: int(x))
df_1["Price"] = df_1.Price.apply(lambda x: float(x))
df_1['Install']= [int(x) if x.isnumeric() else x for x in df_1['Install']]

In [ ]:
#Displaying the datatypes of the dataset

df_1.dtypes

In [ ]:
# IDENTIFYING NUMERICAL FEATURES

numeric_data = df_1.select_dtypes(include=np.number) # select_dtypes selects data with numeric features
numeric_col = numeric_data.columns                                                                

# we will store the numeric features in a variable

print("Numeric Features:")
print(numeric_data.head())

In [ ]:
# IDENTIFYING CATEGORICAL FEATURES

cat_data = df_1.select_dtypes(exclude=np.number) # select_dtypes selects data with numeric features
cat_col = cat_data.columns                                                                

# we will store the numeric features in a variable

print("Categoric Features:")
print(cat_data.head())

### **Visualization - Android Data**

In [ ]:
#Number of categories of apps in the store.
def plot_number_category():
    fig, ax = plt.subplots()
    fig.set_size_inches(15, 7)
    fig.autofmt_xdate()
    countplot=sns.categorical.countplot(df_1.Category,ax=ax)
    plt.xticks(rotation = 90)
    ax.set_title('Number of Categories of Apps in the Store')
    plt.show(countplot)

plot_number_category()

# Tabular representation
top_cat=df_1.groupby('Category').size().reset_index(name='Count').nlargest(5,'Count')
display(top_cat)

The above data shows that the top most category is Education followed by Music&

Audio, Tools, Bussiness and Entertainment.


In [ ]:
top5_cat = top_cat['Category'].tolist()
data_top5 = df_1.groupby('Category')['Install'].agg('sum').loc[top5_cat].reset_index(name='Number_Installations')
data =  df_1.groupby('Category')['Install'].agg('sum').reset_index(name='Number_Installations')

In [ ]:
#Comparing top 5 category on the basis of Installs
fig = plt.figure(figsize=(10,5))
title=plt.title('Comparing top 5 category on the basis of Installs')
bar=sns.barplot(y=data_top5['Category'],x=data_top5['Number_Installations'])
plt.show(bar)

In [ ]:
##Comparing top 5 category on the basis of Content Rating

data_cont =  df_1.groupby('Content Rating')['Install'].agg('sum').reset_index(name='Number_Installations')
plt.figure(figsize=(10,5))
sns.barplot(x=data_cont['Content Rating'], y=data_cont['Number_Installations'])
plt.show()

In [ ]:
# Visualising the Free and Paid applications of android in a pie chart(scale = Percentage)

x= df_1.Free.value_counts()
label= ["Free","Paid"]
mycolors = ["hotpink","black"]
title=plt.title('Free and Paid applications of android')
plt.pie(x,labels= label, autopct= "%1.1f%%", colors = mycolors)
plt.show()

### iOS Dataset

In [ ]:
#importing ios dataset
df_2 = pd.read_json('ios.json')
df_2.head(5)

### Description of iOS Dataset

We are using a json file “ios.json”. Before getting into Actual EDA, we will familiarize with the Variables.

* **App_Id:** This column name represents the application unique Id

* **App_Name:** This column name represents the application name.

* **AppStore_Url:** This column name represents the application Store Url.

* **Primary_Genre:** This coloumn name represents the primary Genre under which the application falls.

* **Content_Rating:** This column name represents the Application content rating.

* **Size_Bytes:** This column name represents the application size (in bytes).

* **Required_IOS_Version:** This column name represents Required IOS version of the Application

* **Released:** This column name represents when the application is released

* **Updated:** This column name represents when the application is updated

* **Version:**This column name represents the application current version

* **Price:** This column name represents the price of the application 

* **Currency:** This column name represents the Currency of the application

* **Free:** This column name represents whether the application is free or not to download

* **DeveloperId:** This column name represents the DeveloperId of the application 

* **Developer:** This column name represents the application Developer name 

* **Developer_Url:** This column name represents the developer Url of the application.

* **Developer_Website:** This column name represents the website of the developer of the application

* **Average_User_Rating:** This column name represents the Average users Rating on the application

* **Reviews:** This column name represents the end users Reviews on application 

* **Current_Version_Score:** This column name represents the Current version score on the application

* **Current_Version_Reviews:** This column name represents the end users Reviews on current version of the application

In [ ]:
# Displaying attributes in datasets

df_2.columns

In [ ]:
# Displaying the statistical summary of the dataframe.

df_2.describe()

In [ ]:
# Displaying the  schema summary of a dataframe

df_2.info()

In [ ]:
# Displaying statistics summary of dataset’s distribution with rows and columns, including NaN values.

df_2.describe(include='all')

In [ ]:
# separating Updated and released columns into 2 different columns (Date and Time seperated).
df_2[['Updated_Date','Updated_Time']] = df_2["Updated"].str.split("T",expand=True)
df_2[['Released_Date','Released_Time']] = df_2["Released"].str.split("T",expand=True)

In [ ]:
# Displaying the dataframe with the updated columns

df_2

In [ ]:
# Dropping the unwanted columns
df_2.drop("AppStore_Url",axis=1, inplace=True)
df_2.drop("Updated_Time",axis=1, inplace=True)
df_2.drop("Released_Time",axis=1, inplace=True)
df_2.drop("Released",axis=1, inplace=True)
df_2.drop("Updated",axis=1, inplace=True)
df_2.drop("DeveloperId",axis=1, inplace=True)
df_2.drop("Developer_Url",axis=1, inplace=True)
df_2.drop("Developer_Website",axis=1, inplace=True)
df_2.drop("Current_Version_Score",axis=1, inplace=True)
df_2.drop("Current_Version_Reviews",axis=1, inplace=True)
df_2.drop("Developer",axis=1, inplace=True)
df_2.drop("Currency",axis=1, inplace=True)

In [ ]:
# Removing the special characters from the Content Rating column 
df_2.Content_Rating = [x.strip().replace('+', '') for x in df_2.Content_Rating]

In [ ]:
# Checking for Null values in the data set
df_2.isnull().sum()

In [ ]:
# Finding the rows in the dataset having bad data

df_2['Content_Rating'].max()

In [ ]:
#Found some bad data with Content Rating saying that Not yet Rated
#Removing the rows with that data
df_2 = df_2[df_2['Content_Rating'] != 'Not yet rated']

In [ ]:
# Verifying if anymore bad data is present in dataset.

df_2['Content_Rating'].max()

In [ ]:
# Check for the datatypes and change if needed

df_2.info()

In [ ]:
# Changing few categorical to numerical
df_2["Content_Rating"]= df_2.Reviews.apply(lambda x: int(x))
df_2["Size_Bytes"]= df_2.Reviews.apply(lambda x: int(x))
df_2["Required_IOS_Version"]= df_2.Reviews.apply(lambda x: float(x))
df_2["Version"]= df_2.Reviews.apply(lambda x: float(x))
df_2["Price"]= df_2.Reviews.apply(lambda x: float(x))
df_2["Updated_Date"] = pd.to_datetime(df_2["Updated_Date"])
df_2["Released_Date"] = pd.to_datetime(df_2["Released_Date"])

In [ ]:
# Displaying the Schema of the dataframe with updated dataset

df_2.info()

### Visualization: iOS Dataset

In [ ]:
#Number of categories of apps in the store.
def plot_number_category():
    fig, ax = plt.subplots()
    fig.set_size_inches(15, 7)
    fig.autofmt_xdate()
    countplot=sns.categorical.countplot(df_2.Primary_Genre,ax=ax)
    plt.xticks(rotation = 90)
    ax.set_title('Number of Categories of Apps in the Store')
    plt.show(countplot)

plot_number_category()

# Tabular representation(Relational format)
top_cat=df_2.groupby('Primary_Genre').size().reset_index(name='Count').nlargest(5,'Count')
display(top_cat)

In [ ]:
# Visualising the Free and Paid applications of iOS in a pie chart(scale = Percentage)

x= df_2.Free.value_counts()
label= ["Free","Paid"]
mycolors = ["hotpink","black"]
plt.pie(x,labels= label,autopct= "%1.1f%%", colors = mycolors)
title=plt.title('Free and Paid applications of iOS')
plt.show()

### Merging Two Datasets

In [ ]:
# Displaying total Available Apps in IOS Dataset

df_2.shape

Total Available Apps in IOS Dataset are 1230366

In [ ]:
# Displaying total Available Apps in Android Dataset

df_1.shape

Total Available Apps in IOS Dataset are 799999

Now , Merge two datasets based on App Name. From that we will get list of apps available in both android and IOS.

In [ ]:
df_3 = pd.merge(df_1, df_2, left_on='App Name', right_on='App_Name')

In [ ]:
# Displaying Total Available Apps in Both Datasets

df_3.shape

Total Available Apps in Both Datasets are 141524.

In [ ]:
#Displaying the attributes from merged dataset

df_3.columns

In [ ]:
### Visualising the distribution of app ratings for Android vs iOS apps

import pandas as pd
import matplotlib.pyplot as plt

# create a subset of the data with only the necessary columns
subset = df_3[["App_Name", "Average_User_Rating"]]

# pivot the data to create a grouped bar chart
pivot_table = subset.pivot_table(index="App_Name", columns=df_3["App_Id"].apply(lambda x: "Android" if x.startswith("com.") else "iOS"), values="Average_User_Rating")
pivot_table.plot(kind="bar", figsize=(10,6))

# set chart title and labels
plt.title("Average User Ratings of Apps on Android and iOS Platforms")
plt.xlabel("App Name")
plt.ylabel("Average User Rating")
plt.legend(title="Platform")
plt.show()